src/Sylius/Bundle/AdminBundle/Controller/ImpersonateUserController.php

```diff
@@ -40,8 +40,8 @@ public function __construct(
         if (null !== $router) {
             trigger_deprecation(
                 'sylius/admin-bundle',
-                '1.4',
-                'Passing a $router as the fourth argument is deprecated and will be prohibited in Sylius 2.0',
+                '1.13',
+                'Not passing a $router as the fourth argument is deprecated and will be prohibited in Sylius 2.0',
```

Comment:
I didn't look further, but there is
```php
if (null !== $router)
```
clause above, which needs to be corrected for the new message as well.

P.S. the way `$this->router` is used in the code, it will crash if it is null.
P.S.S. This logically changes "priority" of referer and generated route usage and referer should become fallback.

All in all it doesn't sound like this deprecation message needs to be updated .

Reply:
You're correct with `if (null !== $router)` 🤝 
but
Correct me if I'm wrong, but if someone doesn't pass the router it will still crash, because there's:
`$this->router->generate(...)`
instead of:
`$this->router?->generate(...)`
So Imo we should force people to pass the router as the mandatory argument or remove the argument and its usage entirely, please share your opinion, because I'm not that sure about this case :)

Reply:
A bit of a strange case, because here the deprecation was thrown out when the router was passed, but at the same time this router was used below without any check. So in this case, looking at the current implementation, the router should be required, otherwise the method below will not work.

Reply:
Message and null-router cases need to be reviewed in separation.

- `1.4` message was correct, I think it was part of initiative to make controller more universal and if impersonation would have been done from grid (Never actually happened), then it will work via `referrer` as it should. So by this logic we should have removed `router` at 2.0.
- They just forgot to update null-router usage.

Merged end result is acceptable, keep `referrer` as first redirect and fallback to customer admin show page, though I'm not sure fallback is needed.

Are there cases, when `referrer` is not set in the wild?